File: Geeks-For-Geeks/000-Mathematics/007.checkPrimes.py
import math
# Naive Approach, check for every number till n, 
# Time Complexity: O(n)

def checkPrimeN(n):
    if n == 1:
        return False
    for i in range(2,n):
        if n % i == 0:
            return False
    return True

print(checkPrimeN(6))


# Efficient approach - Time Complexity: O(sqrt(n))

def checkPrimesEff(n):
    if n == 1:
        return False
    
    for i in range(2,math.ceil(math.sqrt(n))):
        if n % i == 0:
            return False

    return True

print(checkPrimesEff(15))

# Most efficient

def checkPrimes(n):
    if n == 1:
        return False
    if n == 2 or n == 3:
        return True
    if n % 2 == 0 or n % 3 == 0:
        return False
    
    i = 5
    while i * i <= n:
        if n % i == 0 or n % (i+2) == 0:
            return False
        i += 6
    
    return True


print(checkPrimes(21))

File: GeeksForGeeks/DS-Course/000-Mathematics/007.checkPrimes.js
//  Naive Approach, check for every number till n,
//  Time Complexity: O(n)

function checkPrimeN(n) {
  if (n == 1) {
    return false;
  }
  for (let i = 2; i < n; i++) {
    if (n % i == 0) {
      return false;
    }
  }
  return true;
}

// console.log(checkPrimeN(23));

// Efficient approach - Time Complexity: O(sqrt(n))

function checkPrimesEff(n) {
  if (n == 1) {
    return false;
  }
  for (let i = 0; i <= parseInt(Math.sqrt(n)); i++) {
    if (n % i == 0) {
      return false;
    }
  }
  return true;
}

// Most efficient
function checkPrimes(n) {
  if (n == 1) {
    return false;
  }
  if (n == 2 || n == 3) {
    return true;
  }
  if (n % 2 == 0 || n % 3 == 0) {
    return false;
  }
  let i = 5;
  while (i * i <= n) {
    if (n % i == 0 || n % (i + 2) == 0) {
      return false;
    }
    i += 6;
  }
  return true;
}

console.log(checkPrimes(23));
